experimental/dds/tree2/src/simple-tree/proxies.ts

```diff
@@ -1011,6 +1067,46 @@ function extractContentObject<T extends object>(input: T): ExtractedFactoryConte
 	};
 }
 
+/**
+ * If `input` is a factory-created proxy, this will return the content stored in that proxy and forward the `typeNameSymbol` on the proxy to `object`.
+ * Otherwise, it will just return `input`.
+ */
+function extractContentIfProxy<T>(input: T): {
+	content: T;
+	type: "object" | "list" | "map" | "leaf";
+	fromFactory: boolean;
+} {
+	let content = input;
+	let fromFactory = false;
+	const rawEditNode = tryGetEditNode(input);
+	if (rawEditNode !== undefined) {
+		const factoryContent = extractRawNodeContent(rawEditNode);
+		if (factoryContent === undefined) {
+			// We were passed a proxy, but that proxy doesn't have any raw content.
+			throw new Error("Cannot insert a node that is already in the tree");
+		}
+		content = factoryContent as T;
+		fromFactory = true;
+	}
+
+	// TODO: This checking could be made stronger by using `class-tree/Tree.schema(content).kind`
```

Comment:
FYI @CraigMacomber (it did work when I tried it).

Reply:
Can you make this work without a cycle if you directly import treeAPI from class-tree rather than its index file?

Reply:
Since we will be merging the two directories, doing that kind of import to avoid cycles should be fine and is done in a few spots already.

Reply:
It did not like the import to `../class-tree/treeAPi`, but it is okay with imports from `"../class-tree/schemaTypes"`, so I got it working by adding a helper at the bottom of this file (see `getNodeKind`).